### PR TITLE
Integrate Todo, Event and Deadline functionality

### LINKED ISSUE #31: As a user, I want to be able to tag and categorize tasks and events
...so I can organize my schedule more effectively.

### PR DESCRIPTION
**Pull request by clarencepohh**
Closes #31.
Another big PR that modifies lots of classes to fit new functionality. Checkstyle and main tests should pass.

@Mahesh1772 #84 will probably need you to modify large sections of the code in this PR. Do take a closer look at the segments of code that I have partitioned out in TaskManager class to get an idea of what I did. You can change as you see fit to fit your shortened inputs implementation.

@kyhjonathan I had to modify your save and load functions for Storage, in order to accommodate the new task types. Do take a look and modify if you need to.

Let me know if there are any issues.

### CODE REVIEW

File: src/main/java/data/Task.java [hunk 1]
```diff
@@ -76,4 +76,44 @@ public String getStatusIcon() {
...
+    /**
+     * Method that creates the save format for a task.
+     * It is a dummy function meant for the subclass to override.
+     *
+     * @return Empty string.
+     */
+    public String getSaveFormat () {
+        return "";
+    }
 }
```
Comment (NicholasTanYY) [inline src/main/java/data/Task.java:119]: You could consider making these abstract methods instead!

File: src/main/java/data/TaskManager.java [hunk 2]
```diff
@@ -16,31 +16,61 @@
...
+
 import java.util.logging.Level;
 import java.util.logging.Logger;
 
 /**
  * Manages tasks by providing functionalities to add, delete, and update tasks.
  */
 public class TaskManager {
-    private static Logger logger = Logger.getLogger(Logger.GLOBAL_LOGGER_NAME);
+    private static final Logger logger = Logger.getLogger(Logger.GLOBAL_LOGGER_NAME);
```
Comment (NicholasTanYY) [inline src/main/java/data/TaskManager.java:32]: Note the capitalisation in the naming for constants!

**Review by NicholasTanYY — COMMENTED** [review]
Great job on splitting the tasks into different task types! This is a really useful feature for our application! LGTM.

File: src/main/java/data/Task.java [hunk 3]
```diff
@@ -76,4 +76,44 @@ public String getStatusIcon() {
...
+
+    /**
+     * Getter for end date for an Event task.
+     * It is a dummy function meant for the Event subclass to override.
+     *
+     * @return Empty string.
+     */
+    public String getEndDate() {
+        return "";
+    }
```
Comment (Aak242) [inline src/main/java/data/Task.java:98]: These methods can use polymorphism and inherit from a common template, as there are currently several functions that return the same thing

File: src/main/java/data/TaskManager.java [hunk 4]
```diff
@@ -109,22 +143,76 @@ public static void addManager(Scanner scanner, WeekView weekView, boolean inMont
...
+            currentTaskType = EVENT;
+            break;
+
+        default:
+            currentTaskType = null;
+            // intentional fallthrough since addTask method checks for invalid taskType already
+        }
+
+        return currentTaskType;
+    }
```
Comment (Aak242) [inline src/main/java/data/TaskManager.java:205]: Great separation of the parsing logic and the user interface printing for better readability!

File: src/main/java/data/TaskManager.java [hunk 5]
```diff
@@ -163,26 +251,53 @@ public static void updateManager(Scanner scanner, WeekView weekView, boolean inM
...
+                    break;
+
+                case DEADLINE:
+                    String deadline = task.getByDate();
+                    dates[0] = deadline;
+                    break;
+
+                default:
+                    logger.log(Level.INFO, "Task to add was invalid. Task in question: " + taskDescription);
+                }
```
Comment (Aak242) [inline src/main/java/data/TaskManager.java:287]: Good use of logging to log errors or abnormal inputs

File: src/main/java/storage/Storage.java [hunk 6]
```diff
@@ -9,17 +11,21 @@
...
+import static data.TaskManager.addTask;
+import static data.TaskManager.parseTaskType;
+import static data.TaskType.DEADLINE;
+import static data.TaskType.EVENT;
+
 public class Storage {
 
     public static final Path FILE_PATH = Path.of("./save/tasks.txt");
-    private static Logger logger = Logger.getLogger(Logger.GLOBAL_LOGGER_NAME);
+    private static final Logger logger = Logger.getLogger(Logger.GLOBAL_LOGGER_NAME);
```
Comment (Aak242) [inline src/main/java/storage/Storage.java:28]: We should use capitalized names for variables marked as "final" as per the convention

File: src/test/java/data/TaskManagerTest.java [hunk 7]
```diff
@@ -30,27 +29,31 @@ void resetTaskManager() {
...
         // Arrange
         LocalDate date = LocalDate.now();
-        String initialTaskDescription = "Initial task";
-        String updatedTaskDescription = "Updated task";
-        addTask(date, initialTaskDescription);
+        String initialTaskDescription = "Initial todo";
+        String updatedTaskDescription = "Updated todo";
+        TaskType testTaskType = TaskType.TODO;
+        String[] dummyTestDates = new String[]{null};
+        addTask(date, initialTaskDescription, testTaskType, dummyTestDates);
```
Comment (Aak242) [inline src/test/java/data/TaskManagerTest.java:56]: We should consider adding similar testing for Events and Deadlines

**Review by Aak242 — COMMENTED** [review]
Overall, very well-written and managed code, despite the large number of changes that were made. LGTM!

File: src/main/java/data/Task.java [hunk 8]
```diff
@@ -76,4 +76,44 @@ public String getStatusIcon() {
...
+    /**
+     * Getter for by date for a Deadline task.
+     * It is a dummy function meant for the Deadline subclass to override.
+     *
+     * @return Empty string.
+     */
+    public String getByDate() {
+        return "";
+    }
+
```
Comment (Mahesh1772) [inline src/main/java/data/Task.java:109]: If no code is intended inside, consider using abstract methods

File: src/main/java/data/Task.java [hunk 9]
```diff
@@ -76,4 +76,44 @@ public String getStatusIcon() {
...
+    /**
+     * Method that creates the save format for a task.
+     * It is a dummy function meant for the subclass to override.
+     *
+     * @return Empty string.
+     */
+    public String getSaveFormat () {
```
Comment (Mahesh1772) [inline src/main/java/data/Task.java:116]: You can consider saving the name in here, and then overriding the method in the children classes, Deadline and Events

File: src/main/java/data/Deadline.java [hunk 10]
```diff
@@ -16,6 +16,17 @@ public Deadline(String name, String by) {
...
     }
 
+    /**
+     * Getter for by date of Deadline task.
+     * Overrides super dummy function.
+     *
+     * @return The String representation of the by date.
+     */
+    @Override
+    public String getByDate() {
```
Comment (Mahesh1772) [inline src/main/java/data/Deadline.java:26]: You could consider transitioning to LocalDate or LocalDateTime type to make the dates more robust.

File: text-ui-test/input.txt [hunk 11]
```diff
@@ -1,14 +1,26 @@
 add
-18/03/2024
-new task 1
+20/03/2024
+T
+new todo
```
Comment (Mahesh1772) [inline text-ui-test/input.txt:4]: Good job on testing out the new features implemented!

**Review by Mahesh1772 — COMMENTED** [review]
LGTM!

**Pull request merged**